uc-crux-llvm/src/UCCrux/LLVM/Run/Explore.hs

```diff
@@ -79,30 +133,13 @@ explore appCtx modCtx cruxOpts ucOpts halloc =
             (take (Config.exploreBudget ucOpts) (L.modDefines (modCtx ^. llvmModule)))
     let dir = bldDir cruxOpts </> takeFileName (modCtx ^. moduleFilePath) -<.> ""
     createDirectoryIfMissing True dir
+    let funcsToExplore = filter (`notElem` Config.skipFunctions ucOpts) functions
+    let doExplore ac = exploreOne ac modCtx cruxOpts ucOpts halloc dir
     stats <-
-      for (filter (`notElem` Config.skipFunctions ucOpts) functions) $
-        \func ->
-          do
-            let logFilePath = dir </> func -<.> ".summary.log"
-            logExists <- doesPathExist logFilePath
-            if not logExists || Config.reExplore ucOpts
-              then do
-                (appCtx ^. log) Hi $ "Exploring " <> Text.pack func
-                maybeResult <- loopOnFunction appCtx modCtx halloc cruxOpts ucOpts func
-                case maybeResult of
-                  Right (SomeBugfindingResult result) ->
-                    do
-                      writeFile logFilePath (Result.printFunctionSummary (Result.summary result))
-                      pure (getStats result)
-                  Left unin ->
-                    do
-                      writeFile logFilePath (Text.pack (displayException unin))
-                      pure
-                        ( mempty
-                            { unimplementedFreq = Map.singleton (panicComponent unin) 1
-                            }
-                        )
-              else do
-                (appCtx ^. log) Med $ "Skipping already-explored function " <> Text.pack func
-                pure mempty
+      if Config.exploreParallel ucOpts
+        then
+          mapConcurrently
```

Comment:
This could be amusingly dangerous. `mapConcurrently` immediately spawns a thread for each item in the collection. I think you might want something like https://hackage.haskell.org/package/scheduler to bound the number of parallel threads.

Reply:
I've tried this out with a budget as high as 2000 on large projects like `libpng`, `liblzma`, etc. and not had any problems. This is probably in part due to the fact that we don't get very deep into a lot of functions, and so it finishes very quickly for a large percentage of the code.

Anyway, `scheduler` would be an easy addition and could improve performance as well as being less dangerous so I'll integrate it! Thanks for the pointer.

Reply:
Hmm, using the "scheduler" package results in a ~2x slowdown on `libpng` from `mapConcurrently`. What do you think, Tristan? Worth it to avoid problems in the future? Or just keep an eye on it and maybe revisit as the code evolves?

Reply:
This is how I'm using it BTW: https://github.com/GaloisInc/crucible/pull/698/commits/92b4b7624f33d4bd77ac1d6fcbff70e1de7630a1

Reply:
Do you have 16 cores to support that? Passing 0 might be better... it would then just use all available capabilities (i.e., whatever was passed via `+RTS -N`)

Reply:
I do! I also tried with 0/just with the `Par` pattern.

Reply:
I didn't have the right combination of threads in my Cabal file, I wasn't linking the executable with `-threaded`. With that, `scheduler` goes down to 45s, and `mapConcurrently` goes down to *10s*. However, the results from `mapConcurrently` aren't what I'd expect - it seems to be exploring fewer functions that I ask for. It doesn't seem worth time time it would take to debug given that `scheduler` is also the morally correct approach, so I'll call this done for now.